### PR TITLE
Add fPIC options

### PR DESCRIPTION
**Pull request by AlexandreBossard**
useful when you want a pic static library.

### CODE REVIEW

File: conanfile.py [hunk 1]
```diff
@@ -7,7 +7,8 @@
 
 class FmtConan(ConanFile):
     name = "fmt"
-    version = "4.1.0"
+    lib_version = "4.1.0"
+    version = "4.1.0-r1"
```
Comment (uilianries) [inline conanfile.py]: Why did you change the version to r1?

File: conanfile.py [hunk 2]
```diff
@@ -16,19 +17,22 @@ class FmtConan(ConanFile):
     exports_sources = ['CMakeLists.txt']
     generators = 'cmake'
     settings = "os", "compiler", "build_type", "arch"
-    options = {"shared": [True, False], "header_only": [True, False]}
-    default_options = "shared=False", "header_only=False"
+    options = {"shared": [True, False], "header_only": [True, False], "fPIC": [True, False]}
+    default_options = "shared=False", "header_only=False", "fPIC=False"
```
Comment (uilianries) [inline conanfile.py]: Actually we use `fPIC=True` by default, even for static library, you can check in our template: https://github.com/bincrafters/conan-templates/blob/master/conanfile.py#L28

File: conanfile.py [hunk 3]
```diff
@@ -16,19 +17,22 @@ class FmtConan(ConanFile):
...
     source_subfolder = "source_subfolder"
     
     def config_options(self):
         if self.options.header_only:
             self.settings.clear()
             self.options.remove("shared")
+            self.options.remove("fPIC")
+        elif self.options.shared:
```
Comment (uilianries) [inline conanfile.py]: fPIC should not be supported on Windows, so you could replace the line 29 by:

```python
elif self.settings.os == 'Windows':
    del self.options.fPIC
```

File: conanfile.py [hunk 4]
```diff
@@ -37,7 +41,8 @@ def build(self):
             cmake.definitions["FMT_TEST"] = False
             cmake.definitions["FMT_INSTALL"] = True
             cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
-            cmake.definitions["FMT_LIB_DIR"] = "lib"
+            if self.options.fPIC:
```
Comment (uilianries) [inline conanfile.py]: fPIC should not be listed on Windows build, so you could replace the line 44 by:

```python
if self.settings.os != 'Windows':
    cmake.definitions['CMAKE_POSITION_INDEPENDENT_CODE'] = self.options.fPIC
```

**Review by uilianries — CHANGES_REQUESTED** [review]
Hi!

I would like to consider some points:

- fPIC is not usable under Windows platform
- I didn't understand why there is r1 version
- We have a good template [here](https://github.com/bincrafters/conan-templates/blob/master/conanfile.py) where you can take the fPIC support on recipe.

Thanks for your effort!

**Comment by AlexandreBossard** [comment]
Thanks for the review, I'll get to it !

File: conanfile.py (hunk 1 above)
Comment (AlexandreBossard) [reply conanfile.py]: I don't know your policy when you release an updated recipe, but not the actual released version. We append a "release" version here. I guess you lean on the "--outdate" of conan  install ?

**Comment by AlexandreBossard** [comment]
>* fPIC is not usable under Windows platform

Should be fixed, followed the template. 

>* I didn't understand why there is r1 version

Replied to that above, reverted back. Let me know if this is okay.

>* We have a good template here where you can take the fPIC support on recipe.

Was helpful, thanks!

I fixed the libname in the test_package. I believe it was wrong, even if it kind of *worked*.

File: conanfile.py (hunk 1 above)
Comment (uilianries) [reply conanfile.py]: so you are asking for package version, now I see.

We merge all PR over **testing** branch and after to be approved, we merge over **stable**. Only the stable branch is uploaded to Bintray.

 About package version, unfortunately Conan does not support yet and only `self.version` is collected to create the package version. So we don't patch the package version. And how does user know if package is outdated? Well there is `conan install --update` and another command to list outdated packages, but there is no patch number version to checkout if you want to downgrade.

**Comment by uilianries** [comment]
Please, consider to apply the change on discussion https://github.com/bincrafters/conan-fmt/pull/2#discussion_r177189300

**Comment by AlexandreBossard** [comment]
I believe i did, line 33, in the `configure()` function. I followed the template. Is there anything wrong with it ?

**Comment by uilianries** [comment]
My fault, I just didn't find before.

**Review by uilianries — APPROVED** [review]
LGTM

File: conanfile.py (hunk 3 above)
Comment (uilianries) [reply conanfile.py]: Please, remove lines 29 and 30.

Running on Windows, `configure` will remove fPIC option and after that `config_options` will raise an exception trying to remove fPIC. Also, the option value should not be forced, instead, you could raise an exception to warning the user.

**Review by uilianries — CHANGES_REQUESTED** [review]

**Comment by AlexandreBossard** [comment]
I am pretty sure the `config_options()` is run before the `configure()`. I force the `fPIC` options to get the same `package_id` when we have a 'shared' build. Since CMake specify it by default when you build a shared object, and there is no much use of a non `PIC` shared object nowadays.

http://docs.conan.io/en/latest/reference/conanfile/methods.html#configure-config-options
by reading this, I believe the template is wrong, you swapped the definition of `config_options()` and `configure()`.

So I think i've fixed it. I think I have the correct behavior. Specifying either `fPIC` or `shared` options with `header_only` does not change the `package_id`. Specifying a`fPIC` on a windows build will fail, which is good to me, since it does not make sense anyway.

**Comment by uilianries** [comment]
> I am pretty sure the config_options() is run before the configure().

You are right.

>  I force the fPIC options to get the same package_id when we have a 'shared' build. Since CMake specify it by default when you build a shared object, and there is no much use of a non PIC shared object nowadays.

For shared option is okay, I agree.

> http://docs.conan.io/en/latest/reference/conanfile/methods.html#configure-config-options
by reading this, I believe the template is wrong, you swapped the definition of config_options() and configure().

Yes, you are right. I opened an issue to fix it. Thanks for report.

**Review by uilianries — APPROVED** [review]

**Comment by uilianries** [comment]
Many Thanks!
